presto-orc/src/main/java/com/facebook/presto/orc/reader/NullWrappingColumnReader.java

```diff
@@ -0,0 +1,249 @@
+    boolean hasNulls;
+    int innerPosInRowGroup;
+    int numInnerRows;
+    int[] nullsToAdd;
+    int numNullsToAdd;
+    // Number of elements retrieved from inner reader.
+    int numInnerResults;
+
+    protected void beginScan(BooleanInputStream presentStream, LongInputStream lengthStream)
+            throws IOException
+    {
+        super.beginScan(presentStream, lengthStream);
+        numInnerResults = 0;
+    }
+
+    // Translates the positions of inputQualifyingSet between
+    // beginPosition and endPosition into an inner qualifying set for
+    // the non-null content.
+    protected void makeInnerQualifyingSets(int beginPosition, int endPosition)
+    {
+        if (presentStream == null) {
+            numInnerRows = endPosition - beginPosition;
+            hasNulls = false;
+            return;
+        }
+        hasNulls = true;
+        numInnerRows = 0;
+        if (innerQualifyingSet == null) {
+            innerQualifyingSet = new QualifyingSet();
+        }
+        QualifyingSet input = inputQualifyingSet;
+        int[] inputRows = input.getPositions();
+        int numActive = Math.min(endPosition, input.getPositionCount());
+        QualifyingSet inner = innerQualifyingSet;
+        int[] innerRows = inner.getMutablePositions(numActive);
+        int[] innerToOuter = inner.getMutableInputNumbers(numActive);
+        int prevRow = posInRowGroup;
+        int prevInner = innerPosInRowGroup;
+        numNullsToAdd = 0;
+        boolean keepNulls = filter == null || filter.testNull();
+        for (int activeIdx = beginPosition; activeIdx < endPosition; activeIdx++) {
+            int row = inputRows[activeIdx] - posInRowGroup;
+            if (!present[row]) {
+                if (keepNulls) {
+                    addNullToKeep(inputRows[activeIdx]);
+                }
+            }
+            else {
+                int distance = countPresent(prevRow, row);
+                prevRow = row;
+                innerToOuter[numInnerRows] = activeIdx;
+                innerRows[numInnerRows++] = prevInner + distance;
+                prevInner = innerRows[numInnerRows - 1];
+            }
+        }
+        int end = endPosition < inputQualifyingSet.getPositionCount()
+                ? inputRows[endPosition]
+                : input.getEnd();
+        int skip = countPresent(prevRow, end - posInRowGroup);
+        innerQualifyingSet.setPositionCount(numInnerRows);
+        innerQualifyingSet.setEnd(skip + prevInner);
+    }
+
+    private void addNullToKeep(int position)
+    {
+        if (nullsToAdd == null) {
+            nullsToAdd = new int[100];
+        }
+        else if (nullsToAdd.length <= numNullsToAdd) {
+            nullsToAdd = Arrays.copyOf(nullsToAdd, nullsToAdd.length * 2);
+        }
+        nullsToAdd[numNullsToAdd++] = position;
+    }
+
+    protected void shiftUp(int from, int to)
+    {
+        throw new UnsupportedOperationException("subclasses must implement");
```

Comment:
mark this method abstract instead of throwing an exception: protected abstract void shiftUp(int from, int to);